clippy_lints/src/mutable_debug_assertion.rs

```diff
@@ -0,0 +1,105 @@
+    /// between a release and debug build.
+    ///
+    /// **Known problems:** None
+    ///
+    /// **Example:**
+    /// ```rust,ignore
+    /// debug_assert_eq!(vec![3].pop(), Some(3));
+    /// // or
+    /// fn take_a_mut_parameter(_: &mut u32) -> bool { unimplemented!() }
+    /// debug_assert!(take_a_mut_parameter(&mut 5));
+    /// ```
+    pub DEBUG_ASSERT_WITH_MUT_CALL,
+    correctness,
+    "`debug_assert!` and others will be ignored in release mode and therefore alter the execution flow. Replace it with non-mut function calls."
+}
+
+declare_lint_pass!(DebugAssertWithMutCall => [DEBUG_ASSERT_WITH_MUT_CALL]);
+
+fn check_for_mutable_call(cx: &LateContext<'_, '_>, e: &Expr) -> bool {
+    match &e.kind {
+        ExprKind::AddrOf(Mutability::MutMutable, _) => return true,
+        ExprKind::Call(_, args) => return (*args).iter().any(|a| check_for_mutable_call(cx, a)),
+        /* TODO: I can't detect `&mut self`. How to? */
+        ExprKind::MethodCall(_, _, args) => return (*args).iter().any(|a| check_for_mutable_call(cx, a)),
+        ExprKind::Unary(_, ref expr) => return check_for_mutable_call(cx, expr),
+        _ => (),
+    }
+
+    false
+}
+
+fn extract_call(cx: &LateContext<'_, '_>, e: &Expr) -> Option<Span> {
+    if_chain! {
+        if let ExprKind::Block(ref block, _) = e.kind;
+        if block.stmts.len() == 1;
+        if let StmtKind::Semi(ref matchexpr) = block.stmts[0].kind;
+        then {
+            if_chain! {
+                if let ExprKind::Match(ref ifclause, _, _) = matchexpr.kind;
+                if let ExprKind::DropTemps(ref droptmp) = ifclause.kind;
+                if let ExprKind::Unary(UnOp::UnNot, ref condition) = droptmp.kind;
+                then {
+                    // debug_assert
+                    if check_for_mutable_call(cx, condition) {
+                        return Some(condition.span);
+                    }
+                } else {
+                    // debug_assert_{eq,ne}
+                    if_chain! {
+                        if let ExprKind::Block(ref matchblock, _) = matchexpr.kind;
+                        if let Some(ref matchheader) = matchblock.expr;
+                        if let ExprKind::Match(ref headerexpr, _, _) = matchheader.kind;
+                        if let ExprKind::Tup(ref conditions) = headerexpr.kind;
+                        if conditions.len() == 2;
+                        then {
+                            if let ExprKind::AddrOf(_, ref lhs) = conditions[0].kind {
+                                if check_for_mutable_call(cx, lhs) {
+                                    return Some(lhs.span);
+                                }
+                            }
+                            if let ExprKind::AddrOf(_, ref rhs) = conditions[1].kind {
+                                if check_for_mutable_call(cx, rhs) {
+                                    return Some(rhs.span);
+                                }
+                            }
```

Comment:
Can this be simplified further?

Reply:
I'll have a look later. But in the mean time, can you extract this match on `debug_assert` to `utils/higher.rs`, like
https://github.com/rust-lang/rust-clippy/blob/8fae2dd3c1bfed13bdd6c0cfd4170dd1363f25f7/clippy_lints/src/utils/higher.rs#L251

So a function, that matches on `debug_assert{_eq|_ne}` and returns it's arguments. Or even better 2 functions: 1 for the `assert` familiy and one for the `debug_assert` family (reusing the `assert` function)

Reply:
Will do that in the second iteration, but very interesting, yeah
cc #4694